File: src/shared/layout.ts
import { z } from 'zod';

export const windowBoundsSchema = z.object({
  x: z.number(),
  y: z.number(),
  width: z.number(),
  height: z.number(),
});
export type WindowBounds = z.infer<typeof windowBoundsSchema>;

const panelProjectSchema = z.object({ visible: z.boolean(), width: z.number() });
const panelOperationsSchema = z.object({ visible: z.boolean(), height: z.number() });
const panelCellEditorSchema = z.object({ visible: z.boolean(), height: z.number() });

export const layoutSchema = z.object({
  panel: z.object({
    project: panelProjectSchema,
    operations: panelOperationsSchema,
    cellEditor: panelCellEditorSchema,
  }),
  window: z.object({
    bounds: windowBoundsSchema.nullable(),
  }),
});
export type Layout = z.infer<typeof layoutSchema>;

export const layoutPatchSchema = z.object({
  panel: z
    .object({
      project: panelProjectSchema.partial().optional(),
      operations: panelOperationsSchema.partial().optional(),
      cellEditor: panelCellEditorSchema.partial().optional(),
    })
    .optional(),
  window: z
    .object({
      bounds: windowBoundsSchema.nullable().optional(),
    })
    .optional(),
});
export type LayoutPatch = z.infer<typeof layoutPatchSchema>;

export const defaultLayout: Layout = {
  panel: {
    project: { visible: true, width: 260 },
    operations: { visible: false, height: 200 },
    cellEditor: { visible: true, height: 180 },
  },
  window: {
    bounds: null,
  },
};

File: apps/kira-studio/internal/storage/model/layout.go
package model

type PanelProject struct {
	Visible bool    `json:"visible"`
	Width   float64 `json:"width"`
}

type PanelOperations struct {
	Visible bool    `json:"visible"`
	Height  float64 `json:"height"`
}

type PanelCellEditor struct {
	Height float64 `json:"height"`
}

type Layout struct {
	Panel struct {
		Project    PanelProject    `json:"project"`
		Operations PanelOperations `json:"operations"`
		CellEditor PanelCellEditor `json:"cellEditor"`
	} `json:"panel"`
}

// DefaultLayout mirrors packages/shared/domain/layout.ts's defaultLayout verbatim. The window
// rectangle used to live here too (a single `window.bounds` leaf for every window there has ever
// been, P8 F5) — it is now per-window, in the `windows` table (model.WindowRecord, repos.WindowsRepo).
func DefaultLayout() Layout {
	var l Layout
	l.Panel.Project = PanelProject{Visible: true, Width: 260}
	l.Panel.Operations = PanelOperations{Visible: false, Height: 200}
	l.Panel.CellEditor = PanelCellEditor{Height: 180}
	return l
}

// PanelProjectPatch, PanelOperationsPatch and PanelCellEditorPatch mirror layout.ts's
// `.partial()` per-panel patch shapes.
type PanelProjectPatch struct {
	Visible *bool    `json:"visible,omitempty"`
	Width   *float64 `json:"width,omitempty"`
}

type PanelOperationsPatch struct {
	Visible *bool    `json:"visible,omitempty"`
	Height  *float64 `json:"height,omitempty"`
}

type PanelCellEditorPatch struct {
	Height *float64 `json:"height,omitempty"`
}

type PanelsPatch struct {
	Project    *PanelProjectPatch    `json:"project,omitempty"`
	Operations *PanelOperationsPatch `json:"operations,omitempty"`
	CellEditor *PanelCellEditorPatch `json:"cellEditor,omitempty"`
}

type LayoutPatch struct {
	Panel *PanelsPatch `json:"panel,omitempty"`
}
